src/Microsoft.CodeAnalysis.Testing/Microsoft.CodeAnalysis.Analyzer.Testing/Lightup/LightupCompilationWithAnalyzers.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Reflection;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.Testing.Lightup
+{
+    internal static class LightupCompilationWithAnalyzers
+    {
+        private static readonly Type CompilationType = typeof(CompilationWithAnalyzers);
+        private static readonly Assembly FrameworkAssembly = CompilationType.GetTypeInfo().Assembly;
+        private static readonly Type? OptionsType = FrameworkAssembly.DefinedTypes.FirstOrDefault(type => type.Name == "CompilationWithAnalyzersOptions")?.AsType();
+        private static readonly Type? SuppressorType = FrameworkAssembly.DefinedTypes.FirstOrDefault(type => type.Name == "DiagnosticSuppressor")?.AsType();
+        private static readonly int FrameworkMajorVersion = FrameworkAssembly.GetName().Version?.Major ?? 0;
+
+        public static readonly Func<Compilation, ImmutableArray<DiagnosticAnalyzer>, AnalyzerOptions, CancellationToken, CompilationWithAnalyzers> Create = BuildCreatorFunc();
+
+        public static bool IsTestingDiagnosticSuppressors(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        {
+            var suppressorType = SuppressorType;
+
+            if (FrameworkMajorVersion < 4 || suppressorType == null)
+            {
+                return false;
+            }
+
+            return analyzers.Any(analyzer => suppressorType.IsInstanceOfType(analyzer));
+        }
+
+        private static Func<Compilation, ImmutableArray<DiagnosticAnalyzer>, AnalyzerOptions, CancellationToken, CompilationWithAnalyzers> BuildCreatorFunc()
+        {
+            if (OptionsType != null && SuppressorType != null && FrameworkMajorVersion >= 4)
+            {
+                return (compilation, analyzers, options, cancellationToken) =>
+                {
+                    if (!IsTestingDiagnosticSuppressors(analyzers))
+                    {
+                        return compilation.WithAnalyzers(analyzers, options, cancellationToken);
+                    }
+
+                    var compilationWithAnalyzersOptions = Activator.CreateInstance(OptionsType, options, null, true, false, true);
```

Comment:
😕 Can you clarify how this worked in Roslyn 3.x? It seems like this code goes back to Roslyn 1.x, but the tests are currently passing on Roslyn 3.x.

Reply:
Thanks for the explanation!

> Passing a parameter down to the CreateCompilationWithAnalyzers to create two different compilations would break the public API

This isn't necessarily a problem for this repository. We've broken the API several times, and in general this is one of the rarer APIs to override.

Reply:
@tom-englert What if we _always_ request to include suppressed diagnostics, and modify the pragma suppression check code to assert there are no _unsuppressed_ diagnostics present?

Reply:
I'll take a look and see if it works out.

Reply:
➡️ Yes, this was successful. I updated this pull request with the result and extracted the TFM changes and will submit them as a different pull request.